### PR TITLE
Add string type tests

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Targets #182

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Cleanup still missing.

File: src/objects/string.c [hunk 1]
```diff
@@ -274,20 +274,21 @@ ws_string_raw(
...
     ws_object_lock_read(&self->obj);
 
-    if (!self->is_utf8) {
+/*    if (!self->is_utf8) {
+        ws_object_unlock(&self->obj);
         return NULL;
-    }
+    }*/
 
     char* output = calloc(self->charcount + 1, sizeof(*output));
```
Comment (neithernut) [inline src/objects/string.c]: The length should be calles as in `ws_string_set_from_raw()`.

File: tests/check/objects/ws_string/test.c [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    //tcase_add_test(test_string_multicat_empty);
+    tcase_add_test(tc, test_string_empty_dupl);
+    tcase_add_test(tc, test_string_empty_cmp);
+    tcase_add_test(tc, test_string_empty_raw);
+    tcase_add_test(tc, test_string_from_raw_literal);
+    tcase_add_test(tc, test_string_from_raw_heapbuf);
+
+//    tcase_add_test(tc_m, test_string_cmp);
+//    tcase_add_test(tc_m, test_string_ncmp);
+//    tcase_add_test(tc_m, test_string_substr);
```
Comment (neithernut) [inline tests/check/objects/ws_string/test.c]: You realized that I commented these out just to suppress a bit of noise?

Comment (matthiasbeyer) [inline tests/check/objects/ws_string/test.c]: Just noticed you commented out a bunch of stuff... I guess we can investigate together this evening?

**Comment by matthiasbeyer** [comment]
Ready for review here. @neithernut , @TheNeikos

**Comment by neithernut** [comment]
We have to decide what to do about the `utf8` flag. I'd throw it out completely.

**Comment by matthiasbeyer** [comment]
I'm not the one to ask here as I do not really have a clue how this fits all together. We should support UTF-8 in the string type, I really don't care how it is implemented as long as it _is_ implemented.

**Comment by neithernut** [comment]
The flag does nothing but track _whether_ the string is utf8. We can omit it completely and loose nothing. So just rip it out.

**Comment by matthiasbeyer** [comment]
Can we do this in another PR? Would be a bit cleaner!

**Comment by neithernut** [comment]
So you would rather keep commits which jus comment stuff out?

**Comment by matthiasbeyer** [comment]
Nah, I'm a bit fucked up here... I will do it.

**Comment by neithernut** [comment]
ACK for squashing.

**Comment by matthiasbeyer** [comment]
Done. Feel free to suggest more.

File: src/objects/string.c [hunk 3]
```diff
@@ -273,20 +270,16 @@ ws_string_raw(
 ){
     ws_object_lock_read(&self->obj);
 
-    if (!self->is_utf8) {
-        return NULL;
-    }
-
     char* output = calloc(self->charcount + 1, sizeof(*output));
```
Comment (neithernut) [inline src/objects/string.c]: This `calloc()` call has yet to be fixed (see `ws_string_set_from_raw()`).

Comment (matthiasbeyer) [inline src/objects/string.c]: What do you mean exactly?

Comment (neithernut) [inline src/objects/string.c]: The number of units may differ from utf8 to the internal representation.

Comment (matthiasbeyer) [inline src/objects/string.c]: So it should be a `UChar` instead of a `char`? (And a check is missing, wether the `calloc()` worked.=

Comment (neithernut) [inline src/objects/string.c]: No, a pre-flight should be performed, like in `ws_string_set_from_raw()` **just as I pointed out already**.

**Comment by neithernut** [comment]
https://github.com/matthiasbeyer/waysome/commit/0e3f5eb1b43d2d86e3197f0b72cf2e0a13862378 and https://github.com/matthiasbeyer/waysome/commit/53df29d502f7c131287a7678e35c86c4bff17abe should be squased into https://github.com/matthiasbeyer/waysome/commit/136e666c4b51a299d4ce23567215683037b1d8bc.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**